### PR TITLE
feat: add better logging for mqtt

### LINKED ISSUE #406: Better logging for MQTT adapter
#### Reason/Context
When I run the glee app with an `mqtt` broker, I receive the following information in my console:
```
{} Running in development mode...
↙↗ Selected server(s): mqtt
────────────────────────────────────────
⚡ Connected to server mqtt.
```

Although these information are useful, I think it is important to log the channels that have been subscribed to or channels that I can publish to. or if there was an error subscribing to a channel.

#### Expected output:
```
{} Running in development mode...
↙↗ Selected server(s): mqtt
────────────────────────────────────────
⚡ Connected to server mqtt.
⚡ Subscribed to `orders` topic.

```
or 
```
{} Running in development mode...
↙↗ Selected server(s): mqtt
────────────────────────────────────────
⚡ Connected to server mqtt.
x Error while trying to subscribe to `orders` topic.
{Error message}

```
#### Implementation
Add error handling logic to the mqtt Adapter here:
https://github.com/asyncapi/glee/blob/bc0d45f805442c013e6a3a294c533155564d5274/src/adapters/mqtt/index.ts#L138

### PR DESCRIPTION
**Pull request by kaushik-rishi**
**Description**
- Better MQTT logging, for subscription on topics
- More logging suggestions welcome
image
image

**Related issue(s)**
Resolves #406

### CODE REVIEW

**Comment by kaushik-rishi** [comment]
@KhudaDad414 I have a couple of doubts :slightly_smiling_face: 
1. I didn't understand what you mean't by adding error handling logic here https://github.com/asyncapi/glee/blob/master/src/adapters/mqtt/index.ts#L117
2. I did test a failure scenario using some hacky ways, but how can i actually test a failing topic subscription event ?

**Comment by KhudaDad414** [comment]
>I didn't understand what you mean't by adding error handling logic here https://github.com/asyncapi/glee/blob/master/src/adapters/mqtt/index.ts#L117

I don't see any error handing logic there. 🤔 

> I did test a failure scenario using some hacky ways, but how can i actually test a failing topic subscription event ?

have you tried subscribing to a non-existent topic?

**Comment by kaushik-rishi** [comment]
@KhudaDad414 
> I don't see any error handing logic there. :thinking: 

This was regarding the implementation section of this https://github.com/asyncapi/glee/issues/406

You've mentioned to add error handling there, so i asked what it meant :slightly_smiling_face: 

> have you tried subscribing to a non-existent topic?
WDYM by a non existent topic :thinking:  ? The topics are automatically being fetched from the asyncapi.yaml file right ? :slightly_smiling_face:

**Comment by KhudaDad414** [comment]
sorry, I meant here 
https://github.com/asyncapi/glee/blob/bc0d45f805442c013e6a3a294c533155564d5274/src/adapters/mqtt/index.ts#L138

I should have used a permalink instead. 😆 

> WDYM by a non existent topic 🤔 ? The topics are automatically being fetched from the asyncapi.yaml file right ? 🙂

Sorry, I was thinking topics should be declared before subscribing like some other brokers. 😆 
One way that I can think of is trying to subscribe with an invalid `QoS` in binding.

File: src/adapters/mqtt/index.ts [hunk 1]
```diff
@@ -137,6 +138,19 @@ class MqttAdapter extends Adapter {
...
+        if (err) {
+          logLineWithIcon('x', `Error while trying to subscribe to \`${channel}\` topic.`, {
+            highlightedWords: [channel],
+            iconColor: '#f00',
+            disableEmojis: true,
+          })
+          console.log(err.message)
+          return
+        }
+        logLineWithIcon(':zap:', `Subscribed to \`${channel}\` topic.`, {
```
Comment (KhudaDad414) [inline src/adapters/mqtt/index.ts:151]: can you also use that granted object to show the QoS that has been granted?
the message should look something like:
`Subscribed to {topic} topic with QoS {QoS level}.`

Comment (kaushik-rishi) [reply src/adapters/mqtt/index.ts:151]: That should be doable 🙂

Comment (kaushik-rishi) [reply src/adapters/mqtt/index.ts]: Done :slightly_smiling_face:

**Comment by kaushik-rishi** [comment]
@KhudaDad414 
> Sorry, I was thinking topics should be declared before subscribing like some other brokers. laughing One way that I can think of is trying to subscribe with an invalid `QoS` in binding.

I've tested things and they're working as we expected.
image

A couple of things i've discovered while solving this issue, which could help others

1. What's QoS ? (sort of agreement for message delivery between client and broker)
   - At most once (0)
   - At least once (1)
   - Exactly once (2).
2. Invalid binding of qos, was overriden in the library to qos 0, so can't test the error handling using qos
3. To test the error logging, i went through the mqtt package code https://github.com/mqttjs/MQTT.js/blob/main/lib/validations.js
    - this helped me what topic names are invalid

File: src/adapters/mqtt/index.ts [hunk 2]
```diff
@@ -3,6 +3,7 @@ import Adapter from '../../lib/adapter.js'
 import GleeMessage from '../../lib/message.js'
 import { MqttAuthConfig, MqttAdapterConfig } from '../../lib/index.js'
 import { SecurityScheme } from '@asyncapi/parser';
+import { logLineWithIcon } from '../../lib/logger.js';
```
Comment (KhudaDad414) [inline src/adapters/mqtt/index.ts:6]:
```suggestion
import { logLineWithIcon } from '../../lib/logger.js'
```

File: src/adapters/mqtt/index.ts [hunk 3]
```diff
@@ -3,6 +3,7 @@ import Adapter from '../../lib/adapter.js'
 import GleeMessage from '../../lib/message.js'
 import { MqttAuthConfig, MqttAdapterConfig } from '../../lib/index.js'
 import { SecurityScheme } from '@asyncapi/parser';
```
Comment (KhudaDad414) [inline src/adapters/mqtt/index.ts:5]:
```suggestion
import { SecurityScheme } from '@asyncapi/parser'
```

**Review by KhudaDad414 — CHANGES_REQUESTED** [review]
Let's merge this. :)

**Comment by kaushik-rishi** [comment]
@KhudaDad414 Why is this PR not being merged yet, is there any code review that i'm missing out on as this shows "Changes Requested" :thinking: 
I'm kind of confused.

**Review by KhudaDad414 — APPROVED** [review]

**Comment by KhudaDad414** [comment]
Merging it right now. thanks for the contribution. @kaushik-rishi

**Pull request merged**